### PR TITLE
bpo-33972: MIMEPart.iter_attachments blows up more gracefully when co…

### PR DESCRIPTION
**Pull request by stoksc**
https://bugs.python.org/issue33972

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by stoksc** [comment]
CLA is signed now.

File: Lib/email/message.py [hunk 1]
```diff
@@ -1041,7 +1041,10 @@ def iter_attachments(self):
         maintype, subtype = self.get_content_type().split('/')
         if maintype != 'multipart' or subtype == 'alternative':
             return
-        parts = self.get_payload().copy()
+        try:
+            parts = self.get_payload().copy()
+        except AttributeError:
+            raise TypeError('Expected payload as list, got %s' % type(self._payload))
```
Comment (pablogsal) [inline Lib/email/message.py]: Nitpick: Prefer f-strings to do string formatting (or `str.format`, although I prefer f-strings).

File: Lib/test/test_email/test_message.py [hunk 2]
```diff
@@ -798,6 +798,20 @@ def test_set_content_does_not_add_MIME_Version(self):
...
+            Subject: Ayons asperges pour le
+            From: Le Pew <pepe@example.com>
+            To: Penelope Pussycat <penelope@example.com>
+            MIME-Version: 1.0
+            Content-Type: multipart/mixed; charset="utf-8"
+
+            sample text
+            """), policy=policy.default)
+        with self.assertRaises(TypeError) as ar:
+            for a in m.iter_attachments():
```
Comment (pablogsal) [inline Lib/test/test_email/test_message.py]: Maybe substitute this by `next(m.iter_attachments())` to make it more direct.

File: Lib/email/message.py (hunk 1 above)
Comment (pablogsal) [inline Lib/email/message.py]: Use `raise TypeError('Expected payload as list, got %s' % type(self._payload)) from None` to hide the `AttributeError` from the stack trace.

Comment (pablogsal) [inline Lib/email/message.py]: Maybe we can improve the error message taking into account that `The payload will either be a list object or a string`.

**Review by pablogsal — CHANGES_REQUESTED** [review]

**Comment by pablogsal** [comment]
Hi @stoksc and thank you very much for your contribution! The patch looks good, I left just a few issues to address. :)

File: Lib/email/message.py (hunk 1 above)
Comment (stoksc) [reply Lib/email/message.py]: Did you have anything specific in mind?

**Comment by stoksc** [comment]
I have made the requested changes; please review again
Also, thanks for the help @pablogsal! Let me know if you have any other suggestions; your original few were helpful. This is my first time trying to contribute to cpython so thanks for being gentle, ha.

**Pull request closed without merging**